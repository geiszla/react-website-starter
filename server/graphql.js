import {
  GraphQLBoolean,
  GraphQLNonNull,
  GraphQLObjectType,
  GraphQLSchema,
  GraphQLString
} from 'graphql';

import bcrypt from 'bcrypt-nodejs';

// Queries
const queryType = new GraphQLObjectType({
  name: 'Query',
  fields: {
    getUsername: {
      type: GraphQLString,
      resolve: ({ session }) => {
        if (session.isLoggedIn) return session.username;
        return null;
      }
    }
  }
});

// Mutations
const mutationType = new GraphQLObjectType({
  name: 'Mutation',
  fields: {
    loginUser: {
      type: new GraphQLNonNull(GraphQLBoolean),
      args: {
        username: { type: new GraphQLNonNull(GraphQLString) },
        password: { type: new GraphQLNonNull(GraphQLString) }
      },
      resolve: ({ session }, { username, password }) => new Promise((resolve) => {
        const correctPassword = '$2a$08$7OBLs4B/PZvSxafCdjIE8.qost5k7QSyS7tUlgPi0ckDpOYYNAhC.'; // testpass

        bcrypt.compare(password, correctPassword, (err, res) => {
          if (err) {
            console.error('Couldn\'t compare password hases.');
            resolve(false);
          }

          if (res === true) {
            session.isLoggedIn = true;
            session.username = username;

            resolve(session.isLoggedIn);
          } else {
            resolve(false);
          }
        });
      })
    },
    logoutUser: {
      type: new GraphQLNonNull(GraphQLBoolean),
      resolve: ({ session }) => {
        session.isLoggedIn = false;
        return session.isLoggedIn;
      }
    }
  }
});

export default new GraphQLSchema({ query: queryType, mutation: mutationType });
