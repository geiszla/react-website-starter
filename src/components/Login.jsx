import React, { Component } from 'react';

import Button from '@material-ui/core/Button';
import Card from '@material-ui/core/Card';
import CardActions from '@material-ui/core/CardActions';
import CardContent from '@material-ui/core/CardContent';
import PropTypes from 'prop-types';
import TextField from '@material-ui/core/TextField';
import Transition from 'react-transition-group/Transition';
import Typography from '@material-ui/core/Typography';
import { withStyles } from '@material-ui/core/styles';

const styles = () => ({
  card: {
    position: 'relative',
    top: '30%',
    maxWidth: 375,
    margin: 'auto',

    opacity: 0,
    transition: 'opacity 500ms ease-in-out'
  },
  title: {
    fontSize: '20px',
    margin: '7px 0 15px 0'
  },
  button: {
    marginLeft: 'auto'
  }
});

const transitionStyles = {
  entering: { opacity: 0 },
  entered: { opacity: 1 }
};

class Login extends Component {
  handleKeyPress = (event) => {
    if (event.key === 'Enter') {
      this.props.handleLogin();
    }
  }

  render() {
    const {
      classes, usernameError, passwordError, handleLogin
    } = this.props;

    return (
      <Transition appear in timeout={0}>
        {state => (
          <Card className={classes.card} style={{ ...transitionStyles[state] }}>
            <CardContent>
              <Typography type="title" className={classes.title}>
                Choose a name and enter password
              </Typography>
              <TextField
                id="username"
                label="Name"
                margin="dense"
                autoFocus
                error={usernameError}
                onKeyPress={event => this.handleKeyPress(event)}
              />
              <TextField
                id="password"
                label="Password"
                type="password"
                margin="dense"
                autoComplete="current-password"
                error={passwordError}
                onKeyPress={event => this.handleKeyPress(event)}
              />
            </CardContent>
            <CardActions>
              <Button
                variant="raised"
                color="primary"
                className={classes.button}
                onClick={() => handleLogin()}
              >
                Login
              </Button>
            </CardActions>
          </Card>
        )}
      </Transition>
    );
  }
}

Login.propTypes = {
  classes: PropTypes.shape({
    card: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired,
    button: PropTypes.string.isRequired
  }).isRequired,
  handleLogin: PropTypes.func.isRequired,
  usernameError: PropTypes.bool.isRequired,
  passwordError: PropTypes.bool.isRequired
};

export default withStyles(styles)(Login);
