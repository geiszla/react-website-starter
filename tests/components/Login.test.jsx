import { printSchema } from 'graphql';
import { addMockFunctionsToSchema, makeExecutableSchema } from 'graphql-tools';
import React from 'react';
import TestRenderer from 'react-test-renderer';

import schema from '../../server/graphql';
import Login from '../../src/components/Login.jsx';

const executableSchema = makeExecutableSchema({
  typeDefs: printSchema(schema),
  resolverValidationOptions: { requireResolversForResolveType: false }
});

addMockFunctionsToSchema({ schema: executableSchema });

const handleLogin = jest.fn();
const loginRenderer = TestRenderer.create(
  <Login handleLogin={handleLogin} usernameError={false} passwordError={false} />
);

describe('Home component', () => {
  it('should match the snapshot when rendered', () => {
    expect(loginRenderer.toJSON()).toMatchSnapshot();

    const usernameErrorRenderer = TestRenderer.create(
      <Login handleLogin={handleLogin} usernameError passwordError={false} />
    );
    expect(usernameErrorRenderer.toJSON()).toMatchSnapshot();

    const passwordErrorRenderer = TestRenderer.create(
      <Login handleLogin={handleLogin} usernameError={false} passwordError />
    );
    expect(passwordErrorRenderer.toJSON()).toMatchSnapshot();

    const bothErrorRenderer = TestRenderer.create(
      <Login handleLogin={handleLogin} usernameError passwordError />
    );
    expect(bothErrorRenderer.toJSON()).toMatchSnapshot();
  });

  it('should handle key presees in username and password fields', () => {
    const loginCore = loginRenderer.root.find(element => element.type.name === 'Login');

    loginCore.instance.handleKeyPress({});
    expect(handleLogin).not.toHaveBeenCalled();

    loginCore.instance.handleKeyPress({ key: 'Enter' });
    expect(handleLogin).toHaveBeenCalled();
  });

  it('should call handleLogin when login button is pressed', () => {
    const loginButton = loginRenderer.root.find(element => element.type === 'button'
      && element.children
      && element.children.some(child => child.children && child.children[0] === 'Login'));

    loginButton.props.onClick();
    expect(handleLogin).toHaveBeenCalled();
  });
});
