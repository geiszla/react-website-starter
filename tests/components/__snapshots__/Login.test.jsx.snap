// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home component should match the snapshot when rendered 1`] = `
<div
  className="Login-cardContainer-1"
  style={
    Object {
      "opacity": 0,
    }
  }
>
  <div
    className="MuiPaper-root-6 MuiPaper-elevation1-9 MuiPaper-rounded-7 MuiCard-root-5 Login-card-2"
  >
    <div
      className="MuiCardContent-root-33"
    >
      <p
        className="MuiTypography-root-34 MuiTypography-body1-43 Login-title-3"
        type="title"
      >
        Choose a name and enter password
      </p>
      <div
        className="MuiFormControl-root-69 MuiFormControl-marginDense-71"
        onKeyPress={[Function]}
      >
        <label
          className="MuiFormLabel-root-80 MuiInputLabel-root-73 MuiInputLabel-formControl-74 MuiInputLabel-animated-77 MuiInputLabel-marginDense-75"
          data-shrink={false}
          htmlFor="username"
        >
          Name
        </label>
        <div
          className="MuiInputBase-root-100 MuiInput-root-87 MuiInput-underline-91 MuiInputBase-formControl-101 MuiInput-formControl-88 MuiInputBase-marginDense-107"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoFocus={true}
            className="MuiInputBase-input-110 MuiInput-input-95 MuiInputBase-inputMarginDense-111 MuiInput-inputMarginDense-96"
            disabled={false}
            id="username"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            type="text"
          />
        </div>
      </div>
      <div
        className="MuiFormControl-root-69 MuiFormControl-marginDense-71"
        onKeyPress={[Function]}
      >
        <label
          className="MuiFormLabel-root-80 MuiInputLabel-root-73 MuiInputLabel-formControl-74 MuiInputLabel-animated-77 MuiInputLabel-marginDense-75"
          data-shrink={false}
          htmlFor="password"
        >
          Password
        </label>
        <div
          className="MuiInputBase-root-100 MuiInput-root-87 MuiInput-underline-91 MuiInputBase-formControl-101 MuiInput-formControl-88 MuiInputBase-marginDense-107"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoComplete="current-password"
            className="MuiInputBase-input-110 MuiInput-input-95 MuiInputBase-inputType-113 MuiInput-inputType-98 MuiInputBase-inputMarginDense-111 MuiInput-inputMarginDense-96"
            disabled={false}
            id="password"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            type="password"
          />
        </div>
      </div>
    </div>
    <div
      className="MuiCardActions-root-117"
    >
      <button
        className="MuiButtonBase-root-145 MuiButton-root-119 MuiButton-contained-130 MuiButton-containedPrimary-131 MuiButton-raised-133 MuiButton-raisedPrimary-134 Login-button-4 MuiCardActions-action-118"
        disabled={false}
        onBlur={[Function]}
        onClick={[MockFunction]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex="0"
        type="button"
      >
        <span
          className="MuiButton-label-120"
        >
          Login
        </span>
        <span
          className="MuiTouchRipple-root-148"
        />
      </button>
    </div>
  </div>
</div>
`;

exports[`Home component should match the snapshot when rendered 2`] = `
<div
  className="Login-cardContainer-1"
  style={
    Object {
      "opacity": 0,
    }
  }
>
  <div
    className="MuiPaper-root-6 MuiPaper-elevation1-9 MuiPaper-rounded-7 MuiCard-root-5 Login-card-2"
  >
    <div
      className="MuiCardContent-root-33"
    >
      <p
        className="MuiTypography-root-34 MuiTypography-body1-43 Login-title-3"
        type="title"
      >
        Choose a name and enter password
      </p>
      <div
        className="MuiFormControl-root-69 MuiFormControl-marginDense-71"
        onKeyPress={[Function]}
      >
        <label
          className="MuiFormLabel-root-80 MuiFormLabel-error-83 MuiInputLabel-root-73 MuiInputLabel-formControl-74 MuiInputLabel-animated-77 MuiInputLabel-marginDense-75"
          data-shrink={false}
          htmlFor="username"
        >
          Name
        </label>
        <div
          className="MuiInputBase-root-100 MuiInput-root-87 MuiInput-underline-91 MuiInputBase-error-106 MuiInput-error-92 MuiInputBase-formControl-101 MuiInput-formControl-88 MuiInputBase-marginDense-107"
          onClick={[Function]}
        >
          <input
            aria-invalid={true}
            autoFocus={true}
            className="MuiInputBase-input-110 MuiInput-input-95 MuiInputBase-inputMarginDense-111 MuiInput-inputMarginDense-96"
            disabled={false}
            id="username"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            type="text"
          />
        </div>
      </div>
      <div
        className="MuiFormControl-root-69 MuiFormControl-marginDense-71"
        onKeyPress={[Function]}
      >
        <label
          className="MuiFormLabel-root-80 MuiInputLabel-root-73 MuiInputLabel-formControl-74 MuiInputLabel-animated-77 MuiInputLabel-marginDense-75"
          data-shrink={false}
          htmlFor="password"
        >
          Password
        </label>
        <div
          className="MuiInputBase-root-100 MuiInput-root-87 MuiInput-underline-91 MuiInputBase-formControl-101 MuiInput-formControl-88 MuiInputBase-marginDense-107"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoComplete="current-password"
            className="MuiInputBase-input-110 MuiInput-input-95 MuiInputBase-inputType-113 MuiInput-inputType-98 MuiInputBase-inputMarginDense-111 MuiInput-inputMarginDense-96"
            disabled={false}
            id="password"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            type="password"
          />
        </div>
      </div>
    </div>
    <div
      className="MuiCardActions-root-117"
    >
      <button
        className="MuiButtonBase-root-145 MuiButton-root-119 MuiButton-contained-130 MuiButton-containedPrimary-131 MuiButton-raised-133 MuiButton-raisedPrimary-134 Login-button-4 MuiCardActions-action-118"
        disabled={false}
        onBlur={[Function]}
        onClick={[MockFunction]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex="0"
        type="button"
      >
        <span
          className="MuiButton-label-120"
        >
          Login
        </span>
        <span
          className="MuiTouchRipple-root-148"
        />
      </button>
    </div>
  </div>
</div>
`;

exports[`Home component should match the snapshot when rendered 3`] = `
<div
  className="Login-cardContainer-1"
  style={
    Object {
      "opacity": 0,
    }
  }
>
  <div
    className="MuiPaper-root-6 MuiPaper-elevation1-9 MuiPaper-rounded-7 MuiCard-root-5 Login-card-2"
  >
    <div
      className="MuiCardContent-root-33"
    >
      <p
        className="MuiTypography-root-34 MuiTypography-body1-43 Login-title-3"
        type="title"
      >
        Choose a name and enter password
      </p>
      <div
        className="MuiFormControl-root-69 MuiFormControl-marginDense-71"
        onKeyPress={[Function]}
      >
        <label
          className="MuiFormLabel-root-80 MuiInputLabel-root-73 MuiInputLabel-formControl-74 MuiInputLabel-animated-77 MuiInputLabel-marginDense-75"
          data-shrink={false}
          htmlFor="username"
        >
          Name
        </label>
        <div
          className="MuiInputBase-root-100 MuiInput-root-87 MuiInput-underline-91 MuiInputBase-formControl-101 MuiInput-formControl-88 MuiInputBase-marginDense-107"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoFocus={true}
            className="MuiInputBase-input-110 MuiInput-input-95 MuiInputBase-inputMarginDense-111 MuiInput-inputMarginDense-96"
            disabled={false}
            id="username"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            type="text"
          />
        </div>
      </div>
      <div
        className="MuiFormControl-root-69 MuiFormControl-marginDense-71"
        onKeyPress={[Function]}
      >
        <label
          className="MuiFormLabel-root-80 MuiFormLabel-error-83 MuiInputLabel-root-73 MuiInputLabel-formControl-74 MuiInputLabel-animated-77 MuiInputLabel-marginDense-75"
          data-shrink={false}
          htmlFor="password"
        >
          Password
        </label>
        <div
          className="MuiInputBase-root-100 MuiInput-root-87 MuiInput-underline-91 MuiInputBase-error-106 MuiInput-error-92 MuiInputBase-formControl-101 MuiInput-formControl-88 MuiInputBase-marginDense-107"
          onClick={[Function]}
        >
          <input
            aria-invalid={true}
            autoComplete="current-password"
            className="MuiInputBase-input-110 MuiInput-input-95 MuiInputBase-inputType-113 MuiInput-inputType-98 MuiInputBase-inputMarginDense-111 MuiInput-inputMarginDense-96"
            disabled={false}
            id="password"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            type="password"
          />
        </div>
      </div>
    </div>
    <div
      className="MuiCardActions-root-117"
    >
      <button
        className="MuiButtonBase-root-145 MuiButton-root-119 MuiButton-contained-130 MuiButton-containedPrimary-131 MuiButton-raised-133 MuiButton-raisedPrimary-134 Login-button-4 MuiCardActions-action-118"
        disabled={false}
        onBlur={[Function]}
        onClick={[MockFunction]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex="0"
        type="button"
      >
        <span
          className="MuiButton-label-120"
        >
          Login
        </span>
        <span
          className="MuiTouchRipple-root-148"
        />
      </button>
    </div>
  </div>
</div>
`;

exports[`Home component should match the snapshot when rendered 4`] = `
<div
  className="Login-cardContainer-1"
  style={
    Object {
      "opacity": 0,
    }
  }
>
  <div
    className="MuiPaper-root-6 MuiPaper-elevation1-9 MuiPaper-rounded-7 MuiCard-root-5 Login-card-2"
  >
    <div
      className="MuiCardContent-root-33"
    >
      <p
        className="MuiTypography-root-34 MuiTypography-body1-43 Login-title-3"
        type="title"
      >
        Choose a name and enter password
      </p>
      <div
        className="MuiFormControl-root-69 MuiFormControl-marginDense-71"
        onKeyPress={[Function]}
      >
        <label
          className="MuiFormLabel-root-80 MuiFormLabel-error-83 MuiInputLabel-root-73 MuiInputLabel-formControl-74 MuiInputLabel-animated-77 MuiInputLabel-marginDense-75"
          data-shrink={false}
          htmlFor="username"
        >
          Name
        </label>
        <div
          className="MuiInputBase-root-100 MuiInput-root-87 MuiInput-underline-91 MuiInputBase-error-106 MuiInput-error-92 MuiInputBase-formControl-101 MuiInput-formControl-88 MuiInputBase-marginDense-107"
          onClick={[Function]}
        >
          <input
            aria-invalid={true}
            autoFocus={true}
            className="MuiInputBase-input-110 MuiInput-input-95 MuiInputBase-inputMarginDense-111 MuiInput-inputMarginDense-96"
            disabled={false}
            id="username"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            type="text"
          />
        </div>
      </div>
      <div
        className="MuiFormControl-root-69 MuiFormControl-marginDense-71"
        onKeyPress={[Function]}
      >
        <label
          className="MuiFormLabel-root-80 MuiFormLabel-error-83 MuiInputLabel-root-73 MuiInputLabel-formControl-74 MuiInputLabel-animated-77 MuiInputLabel-marginDense-75"
          data-shrink={false}
          htmlFor="password"
        >
          Password
        </label>
        <div
          className="MuiInputBase-root-100 MuiInput-root-87 MuiInput-underline-91 MuiInputBase-error-106 MuiInput-error-92 MuiInputBase-formControl-101 MuiInput-formControl-88 MuiInputBase-marginDense-107"
          onClick={[Function]}
        >
          <input
            aria-invalid={true}
            autoComplete="current-password"
            className="MuiInputBase-input-110 MuiInput-input-95 MuiInputBase-inputType-113 MuiInput-inputType-98 MuiInputBase-inputMarginDense-111 MuiInput-inputMarginDense-96"
            disabled={false}
            id="password"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            type="password"
          />
        </div>
      </div>
    </div>
    <div
      className="MuiCardActions-root-117"
    >
      <button
        className="MuiButtonBase-root-145 MuiButton-root-119 MuiButton-contained-130 MuiButton-containedPrimary-131 MuiButton-raised-133 MuiButton-raisedPrimary-134 Login-button-4 MuiCardActions-action-118"
        disabled={false}
        onBlur={[Function]}
        onClick={[MockFunction]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex="0"
        type="button"
      >
        <span
          className="MuiButton-label-120"
        >
          Login
        </span>
        <span
          className="MuiTouchRipple-root-148"
        />
      </button>
    </div>
  </div>
</div>
`;
